lib/manager/manager.js

```diff
@@ -0,0 +1,317 @@
+ * Copyright (c) 2019, Nodari Chkuaselidze (MIT License)
+ */
+
+'use strict';
+
+const assert = require('bsert');
+const Network = require('bcoin/lib/protocol/network');
+const Logger = require('blgr');
+const EventEmitter = require('events');
+const AbstractDeviceWrapper = require('../device/abstract');
+const LedgerDeviceManager = require('./ledger');
+const TrezorDeviceManager = require('./trezor');
+const MemoryDeviceManager = require('./memory');
+const {AVAILABLE_VENDORS, vendors, parseVendors} = require('../common');
+
+const VENDOR_MANAGERS = {
+  [vendors.LEDGER]: LedgerDeviceManager,
+  [vendors.TREZOR]: TrezorDeviceManager,
+  [vendors.MEMORY]: MemoryDeviceManager
+};
+
+/**
+ * Device Manager
+ *
+ * TODO: Add getDevices.
+ */
+class DeviceManager extends EventEmitter {
+  constructor(options) {
+    super();
+
+    this.opened = false;
+    this.options = new DeviceManagerOptions(options);
+    this.network = this.options.network;
+    this.logger = this.options.logger;
+    this.enabledVendors = this.options.enabledVendors;
+    this.vendorManagers = new Map();
+    this.cachedDevices = new Map();
+    this.selected = null;
+
+    this.init();
+  }
+
+  /**
+   * Initalize device manager,
+   * Create vendor managers.
+   * @private
+   */
+
+  init() {
+    for (const vendor of this.enabledVendors) {
+      const options = this.options.vendorManagerOptions.get(vendor);
+      const manager = new VENDOR_MANAGERS[vendor](options);
+
+      this.vendorManagers.set(vendor, manager);
+    }
+  }
+
+  /**
+   * Start listening to vendor manager events.
+   * @private
+   */
+
+  bind() {
+    const connect = (device) => {
+      const handle = getHandle(device);
+
+      assert(!this.cachedDevices.has(handle),
+        'Already have device for the handle.');
+
+      this.logger.debug('Device connected: ', device);
+
+      this.cachedDevices.set(handle, device);
+      this.emit('connect', device);
+    };
+
+    const disconnect = (device) => {
+      const handle = getHandle(device);
+
+      this.logger.debug('Device disconnected: ', device);
+      this.cachedDevices.delete(handle);
+      this.emit('disconnect', device);
+    };
+
+    const select = (device) => {
+      this.selected = device;
+      this.logger.debug('Device was selected: ', device);
+      this.emit('select', device);
+    };
+
+    const deselect = (device) => {
+      if (device === this.selected) {
+        this.logger.debug('Device was deselected: ', device);
+        this.selected = null;
+      }
+
+      this.emit('deselect', device);
+    };
+
+    for (const manager of this.vendorManagers.values()) {
+      manager.on('connect', connect);
+      manager.on('disconnect', disconnect);
+      manager.on('select', select);
+      manager.on('deselect', deselect);
+    }
+  }
+
+  /**
+   * Select device
+   * @param {String} vendor
+   * @param {DeviceWrapper} device
+   * @returns {Promise<DeviceWrapper>}
+   */
+
+  async selectDevice(vendor, device) {
```

Comment:
Note: I think the device suffix makes sense here, because this class may have other internal properties that can be selected, i.e. potentially wallet clients.